### ISSUE TITLE
Sketch of GPU support

### ISSUE THREAD

**Issue by erictune** [kubernetes/kubernetes#17035]
GPU support:
1. scheduler support (most important)
2. kubelet looks in some flag-configured directory, and sees a file or files that tell it "you have this many of resource example.com/mygpu"
3. scheduler feasibility check that uses any type of resource that any pod or node mentions wanting/having.
4. kubelet support for logical devices, so that containers only see their own GPUs. (less important, but nice to have)
   1. a way to tell kubelet that this set of phsical devices `/dev/nv0` ... `/dev/nv3` are fungible, and map to resource `example.com/mygpu`.
   2. kubelet code to allocate devices, and map them into pods
5. scheduler support for defragging the resource (once we have a rescheduler).

@thockin @davidopp 
requested by @therc

**Comment by erictune** [comment]
Bonus points:

Say you have GPUs which are of different vintages, and some pods want vintage A, and some want vintage B, and others do not care which vintage they get.  For that you need to either have `example.com/my-any-gpu` be equal to the sum of `example.com/my-new-gpu` and `example.com/my-older-gpu` (a derived resource capacity), or have `example.com/my-any-gpu` decrement one resource or the other (but then does kubelet or scheduler decide which was consumed?).  This case is much harder to handle.

**Comment by erictune** [comment]
Hack workaround:

If you only want one GPU pod per machine, you can use a hostPort that represents the GPU resource.

**Comment by davidopp** [comment]
Was @therc requesting merely to schedule at most one GPU-requiring pod per machine (in which case your last comment solves the problem), or something more complicated?

**Comment by erictune** [comment]
He has multiple GPUs per machine and pods with varying GPU count requirements.

**Comment by erictune** [comment]
If we have a generic counted resource, and if the user has  one type of GPU per node, then the user can:

- use a "generic" counted resource for GPU 
- label nodes with the specific GPU flavor 
- pods that need a specific flavor use a node selector
- pods that don't care don't use a node selector

Of course, this makes resource fragmentation, quota, node autoscaling harder.  But it is still an improvement over not having any resource management.

**Comment by erictune** [comment]
@smarterclayton we just talked about how a generic counted resource could be used by a hypothetical customer who is very latency sensitive (e.g. HFT) and cares about NUMA latency too. 

You can either use a generic counted resource for individual cores, or for NUMA-node-sized blocks of cores, and treat that as a counted resource.  You still need to have some code added to kubelet to allocate that counted resource at runtime.

**Comment by davidopp** [comment]
* GPU is simple; I agree with @erictune 's analysis. Though I would say that whether we want to implement a generic counted resource now is debatable. Just doing GPU as a new resource like CPU or memory is a bit easier.

* Reserved cores and NUMA nodes are more complicated because the input from the user is likely to be aggregate CPU or aggregate memory, and then either the scheduler has to know the topology of the machine to know whether something will fit when it's broken down into the schedulable units (cores or NUMA nodes) or it has to be optimistic and then handle the case where the Kubelet rejects because enough aggregate resource is available but not in the chunks needed. Disk is somewhat similar, though disks can be partially filled while reserved cores and NUMA nodes can't be.

So while I think a generic counted resource can definitely be used for GPUs, I don't think it's a good match for cores or NUMA nodes.

BTW @vishh mentioned that the fact the we represent the resources of a container as a map from string to resource quantity (ResourceList is a map[ResourceName]resource.Quantity, to be more precise) should make it easy-ish to do counted resources. It's true that this makes things better than not using a map, but our predicate functions do recognize specific resources (e.g. memory and cpu)
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L152
I guess we could do something like handle all the resources (keys) with specific semantics first, and then handle whatever remains as generic counted resources.

**Comment by quinton-hoole** [comment]
@erictune @therc Can either of you shed any more light on the details of the use case here?  There are several major use cases for GPU's (movie rendering, machine learning...), and they can have substantially different requirements beyond the actual GPU specifics (storage, network bandwidth etc).  We should make sure that we enable the whole use case, not just the GPU part of it.

**Comment by therc** [comment]
This is for machine learning. GPU resource scheduling is, as pointed out, just the tip of the iceberg, but it'd be a start, even in a limited form. An added twist is that, in addition to getting exclusive access to /dev/nvidiaX, the container needs shared access to /dev/nvidiactl and /dev/nvidia-uvm. (I'm trying to not only get something running under Kubernetes, but also tighten the setup)

**Comment by smarterclayton** [comment]
Agree it's more complicated (Numa) but we're looking to enable the bare
minimum support that makes something usable possible for folks who are
already starting to exercise CPU/Numa aware workloads on Kube.  As
discussed with Eric we don't need efficient packing to start, but we do
want to ensure some minimal sanity.  Generic counted (or something like it)
is a good step in that direction if we can do the minimal viable thing.

On Nov 11, 2015, at 1:19 AM, David Oppenheimer <notifications@github.com>
wrote:


   -

   GPU is simple; I agree with @erictune <https://github.com/erictune> 's
   analysis. Though I would say that whether we want to implement a generic
   counted resource now is debatable. Just doing GPU as a new resource like
   CPU or memory is a bit easier.
   -

   Reserved cores and NUMA nodes are more complicated because the input
   from the user is likely to be aggregate CPU or aggregate memory, and then
   either the scheduler has to know the topology of the machine to know
   whether something will fit when it's broken down into the schedulable units
   (cores or NUMA nodes) or it has to be optimistic and then handle the case
   where the Kubelet rejects because enough aggregate resource is available
   but not in the chunks needed. Disk is somewhat similar, though disks can be
   partially filled while reserved cores and NUMA nodes can't be.

So while I think a generic counted resource can definitely be used for
GPUs, I don't think it's a good match for cores or NUMA nodes.

BTW @vishh <https://github.com/vishh> mentioned that the fact the we
represent the resources of a container as a map from string to resource
quantity (ResourceList is a map[ResourceName]resource.Quantity, to be more
precise) should make it easy-ish to do counted resources. It's true that
this makes things better than not using a map, but our predicate functions
do recognize specific resources (e.g. memory and cpu)
https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/scheduler/algorithm/predicates/predicates.go#L152
I guess we could do something like handle all the resources (keys) with
specific semantics first, and then handle whatever remains as generic
counted resources.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/17035#issuecomment-155711642>
.

**Comment by davidopp** [comment]
@smarterclayton Can you say more about the requirements and how you see it working? Is what you have in mind that a user would request some number of NUMA nodes, rather than aggregate memory request? If so, it seems you could instead just have the user request aggregate memory and then have the scheduler round it up to the NUMA node size, without having to introduce a new resource. For example, if the machine has 100 GB of memory capacity, and the NUMA node size is 10 GB, and the machine is already running a pod that requested 72 GB, then a pod that requests 21 GB would not be scheduled there, but a pod that requests 20 would be.

**Comment by jeremyeder** [comment]
The way we are testing HFT workloads now is to key off the PCI ID of the exchange-facing network card/FPGA:


```
# lscpu|grep NUMA
NUMA node(s):          2
NUMA node0 CPU(s):     0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30
NUMA node1 CPU(s):     1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31

# lspci -d 1924:*
41:00.0 Ethernet controller: Solarflare Communications SFC9020 [Solarstorm]
41:00.1 Ethernet controller: Solarflare Communications SFC9020 [Solarstorm]

# for i in `lspci -d 1924:*|awk '{print $1}'` ; do cat /sys/devices/*/*/0000:41:00.0/local_cpulist ; done
1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31
1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31
```

So we know that this adapter is local to node 1.  That means I'd want to schedule my "performance sensitive pod", aka PSP on node 1.  Equivalent to:

```
numactl -N1 -m1 ./my_app.sh
```

In terms of what kernel features need to be surfaced in Kube so that we can encode them in the pod manifest, minimally we'd need 2 things:

1) A way to specify a socket and NUMA memory bank, for cpu-bound apps like anything HPC/energy/DoD etc that may not do a lot of I/O.  SuSE and Red Hat spent a lot of time getting automatic numa_balancing code into the upstream kernel but it does not provide the same guarantees that pinning provides, and it does not yet take into account PCI-locality.

2) A way to specify a PCI device to "follow"
  - In the above example, we'd put device=41:00.0 in the manifest, and Kube would detect (via sysfs) that the device is local to node N.  Then it would write that to the corresponding cpuset.cpus and cpuset.mems files.

Here is an HFT application broken down into what these days we call microservices, and how it would be pinned:

![image](https://cloud.githubusercontent.com/assets/1919640/11161312/39d95de0-8a47-11e5-8f24-e86afcbaf5c4.png)


  http://people.redhat.com/jeder/pics/pinning_map.png

**Comment by jeremyeder** [comment]
@davidopp Field feedback thus far seems to indicate heavily couple with complex sets of nodeSelectors.  Actually here is a question for you all:

If I set schedulable=false on nodeX, and I create a pod with a nodeSelector that only matches nodeX, then I don't think the pod would ever be scheduled -- so perhaps we need a 3rd option for the scheduler:  schedulable=false|true|selected where selected is almost the same as false, except it would accept pods only when it matches a nodeSelector?

**Comment by davidopp** [comment]
> If I set schedulable=false on nodeX, and I create a pod with a nodeSelector that only matches nodeX, then I don't think the pod would ever be scheduled 

Correct.

> so perhaps we need a 3rd option for the scheduler: schedulable=false|true|selected where selected is almost the same as false, except it would accept pods only when it matches a nodeSelector?

Can you explain the use case or this? Are you looking for something like dedicated machines (#17190) ?

**Comment by jeremyeder** [comment]
Yes, thanks.  I think the 3 bullets in the first comment of 17190 would indeed satisfy what we're looking for.  So I will just comment there about that and leave this issue to the tuning options for performance sensitive pods.

As was already noted, the title of this issue is just one use-case amongst many that are very similar.  I hope it can eventually be updated to reflect this.

**Comment by jeremyeder** [comment]
@davidopp @erictune are users free to adjust labels?  I would add area/performance to this one.

**Comment by davidopp** [comment]
Unfortunately the ACL system isn't flexible enough right now for us to be able to let users adjust labels, but I added it for you just now.

**Comment by bgrant0607** [comment]
With respect to scheduling to a specific machine: setting nodeName directly bypasses the scheduler, and so ignores the unschedulable field (deliberately)

**Comment by Hui-Zhi** [comment]
Is there any way to assigned the devices under the cgroup/devices to container by K8s? So far I only know only the devices which has its own cgroup/namespace can be assigned to containers, like CPU/Memory... 
If the devices under cgroup/devices still can't be assigned to container by K8s, I would like to implement it, do there have any task assignment procedure to make me as this task's own?

**Comment by thockin** [comment]
WRT  the any vs old/new - We don't currently report "free" resources on a node, but if/when we do - can it change over time? It has to be possible, right?

Can we define `example.com/new-gpu`, `example.com/old-gpu`, and `example.com/any-gpu` and say that their quantities are not independent variables?  When I allocate an "any" and am granted a "new", the counts of  both "any" and "new" go down.  Yeah, this means there may be scheduling decisions made that are now invalidated (since scheduler doesn't know they are linked) but we have to handle that anyway

**Comment by Hui-Zhi** [comment]
I think all the devices under /sys/fs/cgroup/devices should could be assigned to container in K8s, not just GPU, and so far kubectl doesn't have this ability. I think it should enable it, like Docker "add --device".

**Comment by Hui-Zhi** [comment]
I'm already working on write the code about enable the devices under cgroup/devices. And the changes include the kubectl and scheduler parts.

**Comment by vishh** [comment]
@Hui-Zhi: Can you open a separate issue for devices handling and post your design before you start working on a PR?

**Comment by erictune** [comment]
Yes, please open a issue before you get to deep into coding, @Hui-Zhi

**Comment by erictune** [comment]
@thockin 
I agree you can't have separate resources which count overlapping sets of actual things.

In https://github.com/kubernetes/kubernetes/issues/17035#issuecomment-155526469 I suggested a hacky workaround that gets you part of the way there: a resource for the superset of things, and labels to prefer/requires a subset.

**Comment by Hui-Zhi** [comment]
I'll open a separate issue for it, also will include the design next week earlier.

**Comment by Hui-Zhi** [comment]
I have opened an issue for this,#19049 ,could u help review it?

**Comment by cmluciano** [comment]
Can this issue be closed since multi-GPU support is merged with #42116 ? I have a proposal here for some of the other features mentioned in this issue at kubernetes/community#414
